### PR TITLE
feat: synced preferences

### PR DESCRIPTION
**Pull request by baptiste-grob**
Extracts web preferences to SNJS as a generalized synced preferences mechanism.

Public API is `SNApplication.getPreference()/setPreference()`.

### CODE REVIEW

File: lib/services/preferences_service.ts [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    value: PrefValue[K]
+  ): Promise<void> {
+    if (!this.preferences) return;
+    this.preferences = (await this.itemManager.changeItem<UserPrefsMutator>(
+      this.preferences.uuid,
+      (m) => {
+        m.setPref(key, value);
+      }
+    )) as SNUserPrefs;
+    this.notifyEvent('preferencesChanged');
```
Comment (mobitar) [inline lib/services/preferences_service.ts:72]: Is the string value of this enforced somehow? Would TS complain if you misspelled this?

File: lib/services/singleton_manager.ts [hunk 2]
```diff
@@ -186,73 +187,68 @@ export class SNSingletonManager extends PureService {
...
     createContentType: ContentType,
     createContent: PayloadContent
-  ) {
-    // eslint-disable-next-line no-async-promise-executor
-    return new Promise(async (resolve) => {
-      const matchingItems = this.validItemsMatchingPredicate(predicate);
-      if (matchingItems.length > 0) {
-        return resolve(matchingItems[0]);
-      }
-      /** Item not found, safe to create after full sync has completed */
```
Comment (mobitar) [inline lib/services/singleton_manager.ts:203]: This comment didn't make it in the new code, I think we want to retain it. This logic can be intricate so we want to make sure we have as much context as possible.

File: test/app-group.test.js [hunk 3]
```diff
@@ -88,6 +88,7 @@ describe('application group', () => {
     const identifier = application.identifier;
     await application.deinit(DeinitSource.SignOut);
 
+    await Factory.sleep(0.0001);
```
Comment (mobitar) [inline test/app-group.test.js:91]: Context? Sleep calls in tests should either not be used, be self-evident, or have a comment describing why sleeping is important. Especially if this test passed before with no sleep.

File: test/model_tests/appmodels.test.js [hunk 4]
```diff
@@ -6,7 +6,7 @@ chai.use(chaiAsPromised);
 const expect = chai.expect;
 
 describe('app models', () => {
-  const BASE_ITEM_COUNT = 1; /** Default items key */
+  const BASE_ITEM_COUNT = 2; /** Default items key, user preferences */
```
Comment (mobitar) [inline test/model_tests/appmodels.test.js:9]: I don't remember where, but I started doing this in the style of

```
 const BASE_ITEM_COUNT = ['default-items-key', 'user-preferences'].length;
```

instead to make it self-documenting. Not saying this is necessary here but if we feel the need to refactor this that might be a good way to go.

File: test/preferences.test.js [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    callTimes += 1;
+    await Factory.sleep(0); /** Await next tick */
+    expect(callTimes).to.equal(1); /** App start */
+    await register.call(this);
+    expect(callTimes).to.equal(2);
+    await this.application.setPreference('editorLeft', 300);
+    expect(callTimes).to.equal(3);
+  });
+
+  it('discards existing preferences when logging in', async function () {
```
Comment (mobitar) [inline test/preferences.test.js:82]: Should we also have a test for "retains existing preferences when registering"?

File: test/singletons.test.js [hunk 6]
```diff
@@ -309,6 +309,7 @@ describe('singletons', function() {
       }
     );
     await this.application.modelManager.emitPayload(notErrored);
+    await Factory.sleep(0);
```
Comment (mobitar) [inline test/singletons.test.js:312]: Needs context comment. Is it perhaps to allow decryptErroredItems to finish?

File: test/sync_tests/online.test.js [hunk 7]
```diff
@@ -251,7 +251,7 @@ describe('online syncing', function() {
     // set item to be merged for when sign in occurs
     await this.application.syncService.markAllItemsAsNeedingSync();
     expect(this.application.syncService.isOutOfSync()).to.equal(false);
-    expect(this.application.itemManager.getDirtyItems().length).to.equal(2);
+    expect(this.application.itemManager.getDirtyItems().length).to.equal(3);
```
Comment (mobitar) [inline test/sync_tests/online.test.js:254]: Maybe can be `BASE_ITEM_COUNT + 1`?

**Review by mobitar — APPROVED** [review]

File: lib/services/preferences_service.ts (hunk 1 above)
Comment (baptiste-grob) [reply lib/services/preferences_service.ts:72]:
```ts
class SNPreferencesService extends PureService<'preferencesChanged'>
```
Is what will make TS compain. This is probably not super straightforward but in typescript a string literal can be used as a type. so notifyEvent can only emit events of type `'preferencesChanged'`

File: test/preferences.test.js (hunk 5 above)
Comment (baptiste-grob) [reply test/preferences.test.js:82]: We don't merge preferences at the moment, account preferences are always prioritized.

Comment (mobitar) [reply test/preferences.test.js:82]: Preferences would be auto-merged into registered accounts via the "Merge local items" option on sign in.

**Pull request merged**